openerp/addons/base/ir/ir_attachment.py

```diff
@@ -164,19 +163,37 @@ def _file_write(self, cr, uid, value, checksum):
                 _logger.info("_file_write writing %s", full_path, exc_info=True)
         return fname
 
-    def _file_delete(self, cr, uid, fname):
-        # using SQL to include files hidden through unlink or due to record rules
-        cr.execute("SELECT COUNT(*) FROM ir_attachment WHERE store_fname = %s", (fname,))
-        count = cr.fetchone()[0]
-        full_path = self._full_path(cr, uid, fname)
-        if not count and os.path.exists(full_path):
-            try:
-                os.unlink(full_path)
-            except OSError:
-                _logger.info("_file_delete could not unlink %s", full_path, exc_info=True)
-            except IOError:
-                # Harmless and needed for race conditions
-                _logger.info("_file_delete could not unlink %s", full_path, exc_info=True)
+    def _file_gc(self, cr, uid, context=None):
+        t0 = time.time()
+        # prevent all concurrent updates on ir_attachment while collecting!
+        cr.execute("LOCK ir_attachment IN SHARE MODE")
+        # retrieve the file names to keep
+        cr.execute("SELECT store_fname FROM ir_attachment WHERE store_fname IS NOT NULL")
+        fnames = set(row[0] for row in cr.fetchall())
+        # retrieve all file names, and delete garbage
+        checked = 0
+        removed = 0
+        for (dirpath, _, filenames) in os.walk(self._filestore(cr, uid, context)):
+            dirname = os.path.basename(dirpath)
+            for filename in filenames:
+                checked += 1
+                fname = "%s/%s" % (dirname, filename)
```

Comment:
The `"%s/%s"` is intentional, and follows the convention used by method [_get_path](https://github.com/odoo/odoo/blob/9.0/openerp/addons/base/ir/ir_attachment.py#L128).